runtime/src/serde_utils.rs

```diff
@@ -1,64 +1,670 @@
+        stream,
+        accounts_db,
+        slot,
+        account_storage_entries,
+    )
+}
+
+#[cfg(test)]
+pub(crate) fn accountsdb_from_stream<R, P>(
+    stream: &mut BufReader<R>,
+    account_paths: &[PathBuf],
+    stream_append_vecs_path: P,
+) -> Result<AccountsDB, IoError>
+where
+    R: Read,
+    P: AsRef<Path>,
+{
+    context_accountsdb_from_stream::<DefaultSerdeContext, R, P>(
+        stream,
+        account_paths,
+        stream_append_vecs_path,
+    )
+}
+
+pub fn context_bankrc_from_stream<'a, C, R, P>(
+    account_paths: &[PathBuf],
+    slot: Slot,
+    ancestors: &Ancestors,
+    frozen_account_pubkeys: &[Pubkey],
+    mut stream: &mut BufReader<R>,
+    stream_append_vecs_path: P,
+) -> std::result::Result<BankRc, IoError>
+where
+    C: SerdeContext<'a>,
+    R: Read,
+    P: AsRef<Path>,
+{
+    // read and discard the prepended serialized byte vector length
+    let _len: usize = deserialize_from(&mut stream).map_err(bankrc_to_io_error)?;
+
+    // read and deserialise the accounts database directly from the stream
+    let accounts = context_accounts_from_stream::<C, R, P>(
+        account_paths,
+        ancestors,
+        frozen_account_pubkeys,
+        stream,
+        stream_append_vecs_path,
+    )?;
 
-    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-        formatter.write_str("Expecting bool")
+    Ok(BankRc {
+        accounts: Arc::new(accounts),
+        parent: RwLock::new(None),
+        slot,
+    })
+}
+
+pub fn context_bankrc_to_stream<'a, 'b, C, W>(
+    stream: &'b mut BufWriter<W>,
+    bank_rc: &'a BankRc,
+    snapshot_storages: &'a [SnapshotStorage],
+) -> Result<(), IoError>
+where
+    C: SerdeContext<'a>,
+    W: Write,
+{
+    struct BankRcSerialize<'a, C> {
+        bank_rc: &'a BankRc,
+        snapshot_storages: &'a [SnapshotStorage],
+        phantom: std::marker::PhantomData<C>,
     }
-    fn visit_bool<E>(self, data: bool) -> std::result::Result<bool, E>
-    where
-        E: serde::de::Error,
-    {
-        Ok(data)
+
+    impl<'a, C: SerdeContext<'a>> Serialize for BankRcSerialize<'a, C> {
+        fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+        where
+            S: serde::ser::Serializer,
+        {
+            // first serialize the accounts database to a byte vector
+            let mut buf = Vec::new();
+            context_accountsdb_to_stream::<C, _>(
+                &mut Cursor::new(&mut buf),
+                &*self.bank_rc.accounts.accounts_db,
+                self.bank_rc.slot,
+                self.snapshot_storages,
+            )
+            .map_err(serde::ser::Error::custom)?;
+
+            // then serialize the byte vector to the stream
+            // as compared to serializing directory to the stream,
+            // this effectively prepends a u64 containing the length
+            // of the byte vector to the serialized output
+            serializer.serialize_bytes(&buf)
```

Comment:
My biggest favor and extra bonus review comment is that I really want to get rid of those unneeded bincode indirection.

Actually, `serialize_bytes` isn't needed to use. This just complicates code.See  https://github.com/solana-labs/solana/commit/7bf89d9f26989b667c202d7b1f4edfdcad27c7d0. This clean-up commit is just easy but obviously introduces abi-incompat. So, I really wished it but I couldn't...

(@sakridge: Do you know whether there is subtle reason we forced to use `serialize_bytes` instead of https://github.com/solana-labs/solana/commit/7bf89d9f26989b667c202d7b1f4edfdcad27c7d0?)

Now time has come. :)

If this isn't too much, could you incorporate and correctly versionize this in this pr some way?



This removes these odd code or put versionized code-path at least: 
https://github.com/solana-labs/solana/pull/9980/files#diff-5c68e62b5d516dfc53a73fcbe5e68829R277
https://github.com/solana-labs/solana/pull/9980/files#diff-2099c5256db4eb5975c8834af38f6456L600

Also, this hide abi type info my stall pr #8012..... So I want to fix this in that regard too.

Reply:
As the intent of this was to allow multiple snapshot versions, I've taken out the nested byte array serialisations as part of v1.2.0.

Serde goes out of it's way to ensure you can only serialise one thing and one thing only with the serializer passed into a Serialize::serialize() implementation - so shoving everything into a byte array and serialising that is a crude but very effective workaround for serde's quite intentional limitation in this regard.

It turns out serialising a object as a serde tuple allows you write successive elements out with no prefixed length/size (at least when using the bincode serializer; try this with JSON or some other output format and your mileage may vary). So that's what v1.2.0 does...

Reply:
> As the intent of this was to allow multiple snapshot versions, I've taken out the nested byte array serialisations as part of v1.2.0.

Yeah, I saw this, it's really cool. :)

> Serde goes out of it's way to ensure...

Yeah, that's right. thanks for grokking my comment's intention and serde internals. That's `serde`'s opinionated design decision. You may not like it but let's follow the convention for cleaner code.

Anyway, thanks for taking my big favor and it's really cool that we got one-step closer to the ideal situation!

I think the last thing is to further put the needless serde logic in SerdeContextV1_1_0 and make the common codepath and SerdeContextV1_2_0 really clean.
So, when we add 1_3_0, it'll become dumb easy to copy from ContextV1_2_0 and start coding without worrying about these chaos in the V1_1_0.
To that end, we rather don't want to define `legacy_deserialize_byte_length` for the common trait `SerdeContext`.
Ultimately, when we're 1_5_0 or like, we can safely remove old relics entirely easily.

Lastly, although those well-factored-out common code (like `Serialize for SerializableAccountsDatabaseUnversioned`) looks nice, I anticipate these code is very likely to be affected by newer snapshot changes. In other words, the common code and the actual trait implementation code is so intertwined, making code maintenance difficult a bit.

I think manually writing serialize_tuple is too tedious for SerdeContextV1_2_0+ and i don't want to carry it over SerdeContextV1_3_0...

Well, hard to explain by words, so i've made a sample commit: https://github.com/solana-labs/solana/commit/6d58b73c47294bfb93465d5a83cd2175660b6e6d

Feel free to pick this commit. :)
note: This commit done in hurry. no test updated...

Reply:
@svenski123 Did you read this comment? Sorry I unresolved this thread and added a comment yesterday. :)

Reply:
Ok - I've gone through it all again as well that commit of yours - that's a big commit!

At outset with this PR I was aiming to implement two different but similar serialisation formats with one set of code, reasonably limited changes but structured enough so that I could reason about correctness despite unfamiliarity. To be fair this is not where we are now.

I intentionally avoided making whole new copies of the ser/des logic and the pre and post processing which is non-trivial and imposes a mainentance burden that is undesirable and unnecessary if the old format is retired quickly and there is no need to continue refactoring / developing the new format in the meantime.

If this will take longer to get out there (I suppose 1.1.x to 1.2.x is a hint) and the new format needs to be decoupled more completely from the older one then by all means separate copies is the way as they will diverge - but let's be clear on the requirement and the scope.

Some comments:

- Using serialize_map / serialize_tuple / serialize_seq is actually quite beneficial as it avoids unnecessary copying of data; I've added some generic functions that serialize iterators directly out as seqs, tuples and maps which should hopefully help
- the old snapshot code currently serializes pre-serialized byte sequences but then reads in and discards a u64 and then proceeds to deserialize the data directly rather than deserializing a byte vector and then deserializing that; the point here is the current serde usage is just as broken as serializing a tuple and deserializing indivudal elements and is just as likely (or unlikely) to break in the near term
- I've put the 1.1.0 / legacy specific TypeContext stuff (formerly SerdeContext) stuff in a submodule with the intention that it be jettisoned as soon as it's no longer used
- If having two forks of the ser/des is where we want to go with this - which is fine - then I'd suggest each version go in a separate submodule
- The TypeContext was meant to encapsulate the differences between two ser/des formats - it's kind of ugly in C++, even uglier in Rust and the more the formats differ, the bigger and uglier TypeContext needs to get; and then soon enough it needs to be put down

I see there are some other comments - I'll review and respond there.

Reply:
@svenski123 

> I intentionally avoided making whole new copies of the ser/des logic and the pre and post processing...
> ... let's be clear on the requirement and the scope.

Yeah, I'm sorry for confusing you for requirement and scope.

Considering you started off from the mere omission of count_and_status in AccountStorageEntry, what you original thought as the requirement and scope is completely understandable.

But, your awesome snapshot versioning ground work itself changed the game. That literally broadened the scope for me:
_This work is awesome too good just for the single field omission. Let's make this work the foundation for future snapshot versioning._

Yeah, I didn't communicate it explicitly. Sorry about that.

As a rapidly-developed blockchain project, we wanted a non-breaking versioned snapshot, without worrying too much for temporal code duplication, sacrificing otherwise more well engineered / refactored code.

For context, we did an ad-hoc bank serialization transition like this in the past: https://github.com/solana-labs/solana/pull/9033/files#diff-030d72e3f57f46e47aff0909219268baR35
Yeah, these ad-hoc ness well tolerated here... ;) But, ideally, we wanted to have versioning in place.

So, 

> the new format needs to be decoupled more completely from the older one then by all means separate copies is the way as they will diverge...
> ...
> If having two forks of the ser/des is where we want to go with this - which is fine - then I'd suggest each version go in a separate submodule

I'm afraid but this is true. This is the correct scope/requirement I have now.
Could you rework your PR accordingly each version for each submodule?

PS:

> If this will take longer to get out there (I suppose 1.1.x to 1.2.x is a hint)

I think we can merge this pr this week or early next week. 1.2 branching is almost here. :)

> Using serialize_map / serialize_tuple / serialize_seq is actually quite beneficial as it avoids unnecessary copying of data; I've added some generic functions that serialize iterators directly out as seqs, tuples and maps which should hopefully help

Yeah, I personaly like these gadgets. These will be useful for streaming serialization. But for us, we're selializing on-memory data. So, that's not case... `foo_bar_map.serialize(serializer)` is enough...

> the old snapshot code currently serializes pre-serialized byte sequences but then reads in and discards a u64 and then proceeds to deserialize the data directly rather than deserializing a byte vector and then deserializing that; the point here is the current serde usage is just as broken as serializing a tuple and deserializing indivudal elements and is just as likely (or unlikely) to break in the near term

Yeah, I see. My argument was weak and rather hypothetic. I'm fine to keep your reworked serialization/desealization for old snapshot version in TypeContextLagacy if you're confident. :)

> The TypeContext was meant to encapsulate the differences between two ser/des formats - it's kind of ugly in C++, even uglier in Rust and the more the formats differ, the bigger and uglier TypeContext needs to get; and then soon enough it needs to be put down

Yeah, that's the concern I'm worried with. Our development speed is quite fast and we want to add many changes to Bank/AccountsDB/etc..